### PR TITLE
Adding other option in multiple choice

### PR DESCRIPTION
**Pull request by justinkimtang**
Refactor checkbox recycler

Fixes # [#794](https://github.com/google/ground-android/issues/794)

Updates the multiple choice recyclerView to allow for user input:

- [x] Merges the multiple choice and radio button adapters into one adapter.
- [x] Replaces recyclerview adapter with a list adapter that uses items instead of indexes
- [x] Moves logic for handling data into viewModel
- [x] Uploads user input as an ID field
- [ ] Does not parse the other text or format it in any way 

1) Create a survey with an other option
2) Answer a question with user input in the other survey
3) Submit the survey
4) Verify that the submission on firebase has the other option text set as an Id.

https://github.com/google/ground-android/assets/77077870/a56ee56e-0b62-4b25-86ca-458789343504

@gino-m , @shobhitagarwal1612 PTAL?

### CODE REVIEW

**Comment by shobhitagarwal1612** [comment]
Please verify that the custom value added to other is persisted across sessions. You can test this by killing the app while the data is being collected and then reopen the app. The data collection fragment should resume and should repopulate the previously captured task values.

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2022 Google LLC
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt:2]: Please update the copyright year

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceItem.kt [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package com.google.android.ground.ui.datacollection.tasks.multiplechoice
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceItem.kt:1]: Please add license header

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt [hunk 3]
```diff
@@ -16,15 +16,108 @@
...
+        MultipleChoiceItem(
+          option,
+          isOptionSelected(option),
+          if (isMultipleChoice) MultipleChoiceItem.Type.CHECKBOX else MultipleChoiceItem.Type.RADIO,
+        )
+      }
+    )
+
+    if (multipleChoice.hasOtherOption) {
+      val otherOption = Option(otherId, "", "Other")
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt:64]: Please move the user-visible text to strings.xml

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+ * checkbox views.
+ */
+class MultipleChoiceAdapter(private val viewModel: MultipleChoiceTaskViewModel) :
+  ListAdapter<MultipleChoiceItem, RecyclerView.ViewHolder>(MultipleChoiceItemCallback()) {
+
+  override fun getItemViewType(position: Int): Int {
+    return when (getItem(position).type) {
+      MultipleChoiceItem.Type.CHECKBOX -> R.layout.multiple_choice_checkbox_item
+      else -> R.layout.multiple_choice_radiobutton_item
+    }
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt:41]: Instead of using layout id as item type, can we add an integer value to the enum `MultipleChoiceItem.Type`?

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+          )
+        )
+    }
+
+  override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
+    if (holder is CheckBoxViewHolder) {
+      holder.bind(getItem(position), viewModel)
+    } else if (holder is RadioButtonViewHolder) {
+      holder.bind(getItem(position), viewModel)
+    }
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt:69]: Can we create an abstract ViewHolder which contains the `bind` method?

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskFragment.kt [hunk 6]
```diff
@@ -45,36 +44,15 @@ class MultipleChoiceTaskFragment : AbstractTaskFragment<MultipleChoiceTaskViewMo
...
-    if (isMultipleChoice) {
-      SelectMultipleOptionAdapter(options, selectedIndices) { updateResponse(it) }
-    } else {
-      assert(selectedIndices.size < 2) {
-        "Expected size to be less than 2, found ${selectedIndices.size}"
-      }
-      val selectedIndex = if (selectedIndices.size == 1) selectedIndices[0] else -1
-      SelectOneOptionAdapter(options, selectedIndex) { updateResponse(listOf(it)) }
-    }
+    viewModel.itemsLivedata.observe(this) { updateAdapterList(it) }
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskFragment.kt:52]: Can we replace LiveData with flow?

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt [hunk 7]
```diff
@@ -16,15 +16,108 @@
...
+    }
+    backingItems.value = items
+  }
+
+  private fun isOptionSelected(option: Option): Boolean {
+    return selectedIds.contains(option.id)
+  }
+
+  fun onOptionClicked(item: MultipleChoiceItem) {
+    val isSelected = !isOptionSelected(item.option)
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt:82]: Should it be `val isSelected = isOptionSelected(item.option)` ? (without negation)

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt [hunk 8]
```diff
@@ -16,15 +16,108 @@
...
+    if (item.type == MultipleChoiceItem.Type.RADIO) {
+      selectedIds.clear()
+      if (isSelected) selectedIds.add(item.option.id)
+    } else {
+      if (isSelected) {
+        selectedIds.add(item.option.id)
+      } else {
+        selectedIds.remove(item.option.id)
+      }
+    }
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt:92]: Multiple levels of nesting is making it hard to follow the code. Can we flatten it somehow?

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt [hunk 9]
```diff
@@ -16,15 +16,108 @@
...
+        task.multipleChoice,
+        selectedIds.map { if (it == otherId) formatOtherText(otherText) else it },
+      )
+    )
+    updateMultipleChoiceItems()
+  }
+
+  private fun formatOtherText(otherText: String): String {
+    return "[ $otherText ]"
+  }
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt:108]: Is this formatting applied to other values as well? Or is it a product requirement for "other" type?

File: ground/src/main/res/layout/multiple_choice_checkbox_item.xml [hunk 10]
```diff
@@ -1,32 +1,50 @@
...
-  ~
-  ~ Unless required by applicable law or agreed to in writing, software
-  ~ distributed under the License is distributed on an "AS IS" BASIS,
-  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-  ~ See the License for the specific language governing permissions and
-  ~ limitations under the License.
-  -->
-
-<layout xmlns:android="http://schemas.android.com/apk/res/android">
+<layout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (shobhitagarwal1612) [inline ground/src/main/res/layout/multiple_choice_checkbox_item.xml:3]: Please add license header

**Review by shobhitagarwal1612 — CHANGES_REQUESTED** [review]

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt (hunk 4 above)
Comment (justinkimtang) [reply ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt]: I opted for just passing in a boolean into the adapter which specifies its type.

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt (hunk 5 above)
Comment (justinkimtang) [reply ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt]: Done.

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt (hunk 7 above)
Comment (justinkimtang) [reply ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt]: Not in this case. Here we are toggling the selection. So we get the old selected state and take the neagtion of it.
I'll update this to be clearer!

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt (hunk 9 above)
Comment (justinkimtang) [reply ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt]: This is done just for the other type. @gino-m can give more info.
But this is a hacky way to have android upload user answers without updating the backend to accept the new value. We are basically just making up an ID that contains the user created text.

File: ground/src/main/res/layout/multiple_choice_checkbox_item.xml (hunk 10 above)
Comment (justinkimtang) [reply ground/src/main/res/layout/multiple_choice_checkbox_item.xml:19]: Done.

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskFragment.kt (hunk 6 above)
Comment (justinkimtang) [reply ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskFragment.kt]: Done.

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+class MultipleChoiceAdapter(
+  private val viewModel: MultipleChoiceTaskViewModel,
+  private val canSelectMultiple: Boolean,
+) : ListAdapter<MultipleChoiceItem, RecyclerView.ViewHolder>(MultipleChoiceItemCallback) {
+
+  override fun getItemViewType(position: Int): Int {
+    return when (canSelectMultiple) {
+      true -> R.layout.multiple_choice_checkbox_item
+      else -> R.layout.multiple_choice_radiobutton_item
+    }
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt:39]: This still doesn't resolve the original question. IMO, we shouldn't reuse the layout ids as item type. It should be some integer constant.

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder =
+    when (viewType) {
+      R.layout.multiple_choice_checkbox_item ->
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt:44]: Same as above

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskFragment.kt [hunk 13]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2022 Google LLC
+ * Copyright 2024 Google LLC
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskFragment.kt:2]: We shouldn't update the year for an existing class

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt [hunk 14]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2020 Google LLC
+ * Copyright 2024 Google LLC
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt:2]: Same as above

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt [hunk 15]
```diff
@@ -16,15 +16,111 @@
...
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.filterNotNull
 
 class MultipleChoiceTaskViewModel @Inject constructor(resources: Resources) :
   AbstractTaskViewModel(resources) {
+  companion object {
+    private const val otherId: String = "OTHER_ID"
+    private const val otherPrefix: String = "[ "
+    private const val otherSuffix: String = " ]"
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt:39]: Please use upper casing for private const variables

File: ground/src/main/res/layout/multiple_choice_checkbox_item.xml [hunk 16]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 
 <!--
-  ~ Copyright 2022 Google LLC
+  ~ Copyright 2024 Google LLC
```
Comment (shobhitagarwal1612) [inline ground/src/main/res/layout/multiple_choice_checkbox_item.xml:4]: Same as previous comment

File: ground/src/main/res/layout/multiple_choice_radiobutton_item.xml [hunk 17]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 
 <!--
-  ~ Copyright 2022 Google LLC
+  ~ Copyright 2024 Google LLC
```
Comment (shobhitagarwal1612) [inline ground/src/main/res/layout/multiple_choice_radiobutton_item.xml:4]: Same as previous comment

**Review by shobhitagarwal1612 — CHANGES_REQUESTED** [review]

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt (hunk 11 above)
Comment (gino-m) [reply ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt:39]: Also, I think if / else would be more idiomatic here?

**Review by gino-m — CHANGES_REQUESTED** [review]
Nice work, @justinkimtang! Please run `gradle checkCode` to see failing detekt style checks. Besides that and @shobhitagarwal1612 's comments LGTM!

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt (hunk 11 above)
Comment (justinkimtang) [reply ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceAdapter.kt]: Sorry I misunderstood the first comment, fixing!

**Review by shobhitagarwal1612 — APPROVED** [review]
Thanks! LGTM

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt [hunk 18]
```diff
@@ -16,15 +16,116 @@
...
+    updateMultipleChoiceItems()
+  }
+
+  fun updateResponse() {
+    setValue(
+      fromList(
+        task.multipleChoice,
+        selectedIds.map {
+          // if the other option is selected grab the associated text save instead of otherId
+          if (it == OTHER_ID) OTHER_PREFIX + otherText.trim() + OTHER_SUFFIX else it
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt:86]: This workaround sounds like something that would be specific to the remote persistence layer, not the view. Instead, I would expect the "other" text to be set in the `otherText` field we added to the model, and our ad hoc encoding only represented in one of the remote persistence converters . Can we do that as a fast-follow?

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt [hunk 19]
```diff
@@ -16,15 +16,116 @@
...
+
+  /* Reads the saved task value and adds selected items to the selected list*/
+  private fun loadPendingSelections() {
+    val selectedOptionIds = (taskValue.value as? MultipleChoiceResponse)?.selectedOptionIds
+    val multipleChoice = checkNotNull(task.multipleChoice)
+    val optionIds = multipleChoice.options.map { option -> option.id }
+    selectedOptionIds?.forEach {
+      // if the saved option id was the user inputted other text save the text to other text and add
+      // other id to selected ids
+      if (!optionIds.contains(it)) {
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt:119]: Same here

**Review by gino-m — APPROVED** [review]
Created https://github.com/google/ground-android/issues/2343 to track follow-up. Thanks for the fix!

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt (hunk 18 above)
Comment (justinkimtang) [reply ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt:86]: Yeah makes sense, lets make a bug for it.
